crates/wasm-mutate/src/mutators/peephole/eggsy/expr_enumerator.rs

```diff
@@ -0,0 +1,877 @@
+        let lcope = $left;
+        let rcopy = $right;
+        // Cartesian product of the operands
+        let t = lazy_expand(lcope, $egraph.clone(), $depth, $rnd, &$recexpr)
+            .flat_map(move |e| {
+                std::iter::repeat(e).zip(lazy_expand(
+                    rcopy,
+                    $egraph.clone(),
+                    $depth,
+                    $rnd,
+                    &$recexpr,
+                ))
+            })
+            .map(move |(l, r)| $recexpr.borrow_mut().add(Lang::$lang([l, r])));
+
+        Box::new(t)
+    }};
+}
+
+macro_rules! unop {
+    ($lang:ident,$arg: ident, $egraph: ident, $rnd: ident, $depth: ident, $recexpr: ident) => {{
+        let lcope = $arg;
+        let t = lazy_expand(lcope, $egraph, $depth, $rnd, &$recexpr)
+            .map(move |l| $recexpr.borrow_mut().add(Lang::$lang([l])));
+
+        Box::new(t)
+    }};
+}
+
+macro_rules! store {
+    ($lang:ident, $left:ident, $right: ident, $egraph: ident, $rnd: ident, $depth: ident, $recexpr: ident, $align: ident, $offset: ident, $mem: ident) => {{
+        let lcope = $left;
+        let rcopy = $right;
+        let t = lazy_expand(lcope, $egraph.clone(), $depth, $rnd, &$recexpr)
+            .flat_map(move |e| {
+                std::iter::repeat(e).zip(lazy_expand(
+                    rcopy,
+                    $egraph.clone(),
+                    $depth,
+                    $rnd,
+                    &$recexpr,
+                ))
+            })
+            .map(move |(l, r)| {
+                $recexpr.borrow_mut().add(Lang::$lang {
+                    value_and_offset: [l, r],
+                    align: $align,
+                    mem: $mem,
+                    static_offset: $offset,
+                })
+            });
+
+        Box::new(t)
+    }};
+}
+
+macro_rules! load {
+    ($lang:ident, $arg: ident, $egraph: ident, $rnd: ident, $depth: ident, $recexpr: ident, $align: ident, $offset: ident, $mem: ident) => {{
+        let lcope = $arg;
+        let t = lazy_expand(lcope, $egraph, $depth, $rnd, &$recexpr).map(move |l| {
+            $recexpr.borrow_mut().add(Lang::$lang {
+                offset: l,
+                static_offset: $offset,
+                align: $align,
+                mem: $mem,
+            })
+        });
+
+        Box::new(t)
+    }};
+}
+
+macro_rules! local_or_global {
+    ($lang:ident, $idx: ident, $arg: ident, $egraph: ident, $rnd: ident, $depth: ident, $recexpr: ident) => {{
+        let lcope = $arg;
+        let t = lazy_expand(lcope, $egraph, $depth, $rnd, &$recexpr)
+            .map(move |l| $recexpr.borrow_mut().add(Lang::$lang($idx, l)));
+
+        Box::new(t)
+    }};
+}
+
+/// Returns a lazy iterator over random expressions in the eggraph
+///
+pub fn lazy_expand<'a>(
+    id: Id,
+    egraph: EG,
+    depth: u32,
+    rnd: &'a RefCell<&'a mut SmallRng>,
+    recexpr: &'a RefCell<RecExpr<Lang>>,
```

Comment:
It isn't clear to me why this needs to be in a `RefCell` as well.

In general, we should try and avoid `RefCell`s unless there is a really good reason, as it turns compile-time guarantees to run-time checks.

Reply:
Comments above